templates/deployments/appname/app.jsonnet.tpl

```diff
@@ -129,6 +159,69 @@ local all = {
 			'fflags.yaml': std.manifestYamlDoc(this.data_),
 		},
 	},
+{{- if eq "canary" (stencil.Arg "deployment.strategy") }}
+	service_canary: ok.Service(app.name + '-canary', app.namespace) {
+		target_pod:: $.deployment.spec.template,
+		metadata+: {
+			labels+: sharedLabels,
+		},
+		spec+: $.service.spec,
+	},
+	service_stable: ok.Service(app.name + '-stable', app.namespace) {
+		target_pod:: $.deployment.spec.template,
+		metadata+: {
+			labels+: sharedLabels,
+		},
+		spec+: $.service.spec
+	},
+	analysis_template: argo.AnalysisTemplate('metrics', app) {
+		metrics:: deploymentMetrics
+	},
+	canary_deployment: argo.CanaryDeployment(app.name, app.namespace) {
+		deploymentRef:: $.deployment,
+		canaryService:: $.service_canary,
+		stableService:: $.service_stable,
+		steps:: (if !isDev then [
+			{ setWeight: 25 },
+			{ pause: { duration: '5m' } },
+			{ setWeight: 50 },
+			{ pause: { duration: '5m' } },
+			{ setWeight: 75 },
+			{ pause: { duration: '5m' } },
+		] else []) + [
+			{ setWeight: 100 },
+		],
+		{{- $servicePort := "" }}
+		{{- if (has "http" (stencil.Arg "serviceActivities")) }}
+		{{- $servicePort = 8080 }}
+		{{- end }}
+		{{- if (has "grpc" (stencil.Arg "serviceActivities")) }}
+		{{- $servicePort = 5000 }}
+		{{- end }}
+		{{- if $servicePort }}
+		servicePort:: {{ $servicePort }},
+		{{- end }}
+		{{- if stencil.Arg "slack" }}
+		notification_success:: {{ stencil.Arg "slack" | squote }},
+		notification_failure:: {{ stencil.Arg "slack" | squote }},
+		{{- end }}
+		backgroundAnalysis:: if !isDev then {
+			templates: [
+				argo.AnalysisTemplateRef($.analysis_template),
+			],
+			startingStep: 2,
+		},
+		metadata+: {
+			labels+: sharedLabels,
+			annotations+: {
+				'link.argocd.argoproj.io/external-link': 'https://argorollouts.%(bento)s.%(region)s.outreach.cloud/rollouts/rollout/%(namespace)s/%(name)s' % app,
+			},
+		},
+		spec+: {
+			replicas: if isDev then 1 else 2,
```

Comment:
Im thinking I might be safer to let users manually do it thru override after their own validation.

Reply:
Makes sense, will this cause any conflicts because two things would own it, right?

Reply:
Between Deployment and Rollout? I do not think it would cause any conflicts. per my understanding Rollouts.WorkloadRef takes only pod template spec from Deployment.